java/src/main/java/ai/rapids/cudf/ORCChunkedReader.java

```diff
@@ -0,0 +1,164 @@
+/*
+ *
+ *  Copyright (c) 2024, NVIDIA CORPORATION.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ */
+
+package ai.rapids.cudf;
+
+/**
+ * Provide an interface for reading an ORC file in an iterative manner.
+ */
+public class ORCChunkedReader implements AutoCloseable {
+  static {
+    NativeDepsLoader.loadNativeDeps();
+  }
+
+  /**
+   * Construct the reader instance from read limits, output row granularity,
+   * and a file already loaded in a memory buffer.
+   *
+   * @param chunkReadLimit Limit on total number of bytes to be returned per read,
+   *                       or 0 if there is no limit.
+   * @param passReadLimit  Limit on the amount of memory used by the chunked reader,
+   *                       or 0 if there is no limit.
+   * @param opts           The options for ORC reading.
+   * @param buffer         Raw ORC file content.
+   * @param offset         The starting offset into buffer.
+   * @param len            The number of bytes to parse the given buffer.
+   */
+  public ORCChunkedReader(long chunkReadLimit, long passReadLimit,
+      ORCOptions opts, HostMemoryBuffer buffer, long offset, long len) {
+    handle = createReader(chunkReadLimit, passReadLimit,
+        opts.getIncludeColumnNames(), buffer.getAddress() + offset, len,
+        opts.usingNumPyTypes(), opts.timeUnit().typeId.getNativeId(),
+        opts.getDecimal128Columns());
+    if (handle == 0) {
+      throw new IllegalStateException("Cannot create native chunked ORC reader object.");
+    }
+  }
+
+  /**
+   * Construct a chunked ORC reader instance, similar to
+   * {@link ORCChunkedReader#ORCChunkedReader(long, long, ORCOptions, HostMemoryBuffer, long, long)},
+   * with an additional parameter to control the granularity of the output table.
+   *
+   * @param outputRowGranularity The change step in number of rows in the output table.
```

Comment:
Not sure what is meant by "change step" here.  Is this a maximum number of rows per output table or something else?

Reply:
This mean size (num rows) the output table will be adjusted by this step, such that its size bytes will be closes to the chunk output limit. For example, given granularity is 10k rows (default value), the output table will have either (`10k/20k/30k.... + n`) rows, where `n` is some reminder number.

This granularity is to mitigate the overhead of computing size bytes of every row. By computing size bytes of 10k rows instead, we can save some computation time. The smaller granularity, the more overhead but the output table is closer to the given output limit.

That docs is already in the C++ ORC chunked reader. If you think it worth copying here then I'll do that.

Reply:
Ah, yeah so maybe rowSizingGranularity is more apropos?  Would be good to have some more details here because as-is I don't think readers will understand what this parameter does without having to spelunk all the way through to the C++ code to see how it's used.

Reply:
@jlowe This overload will be unused. The plugin will just call the version without such granularity parameter (which internally uses 10k rows granularity). I added this overload just in case we want to adjust such granularity in the plugin in the future.